tests/integration/test_random_creation.py

```diff
@@ -33,6 +35,366 @@ def test_randn():
+    arr_num = np.random.rand(*shape)
+    assert arr_np.__class__.__name__ == arr_num.__class__.__name__
+    if hasattr(arr_np, "__array__"):
+        assert arr_np.shape == arr_num.shape
+        assert_distribution(
+            arr_num, np.mean(arr_np), np.std(arr_np), mean_tol=0.05
+        )
+
+
+LOW_HIGH = [
+    (5, 10000),
+    (-10000, 5),
+    pytest.param(
+        3000.45,
+        15000,
+        marks=pytest.mark.xfail(
+            reason="NumPy pass, cuNumeric hit struct.error"
+        ),
+        # When size is greater than 1024, legate core throws error
+        # struct.error: required argument is not an integer
+    ),
+    (10000, None),
+]
+SIZES = [
+    10000,
+    pytest.param(
+        None,
+        marks=pytest.mark.xfail(
+            reason="NumPy returns scalar, cuNumeric returns 1-dim array"
+            # with dtype:
+            # np.random.randint(5, 10000, None, np.int64).ndim  << 0
+            # num.random.randint(5, 10000, None, np.int64).ndim << 1
+            # without dtype:
+            # np.random.randint(5, 10000, None).__class__
+            #     <class 'int'>
+            # num.random.randint(5, 10000, None).__class__
+            #     <class 'cunumeric.array.ndarray'>
+        ),
+    ),
+    0,
+    (20, 50, 4),
+]
+
+
+@pytest.mark.parametrize("low, high", LOW_HIGH, ids=str)
+@pytest.mark.parametrize("size", SIZES, ids=str)
+@pytest.mark.parametrize(
+    "dtype",
+    [
+        np.int64,
+        pytest.param(
+            np.uint16,
+            marks=pytest.mark.xfail(
+                reason="cuNumeric raises NotImplementedError"
+            ),
+        ),
+    ],
+    ids=str,
+)
+def test_randint(low, high, size, dtype):
+    arr_np, arr_num = gen_random_from_both(
+        "randint", low=low, high=high, size=size, dtype=dtype
+    )
+    assert arr_np.shape == arr_num.shape
+    # skip distribution assert for small arrays
+    if arr_np.size > 1024:
+        assert_distribution(
+            arr_num, np.mean(arr_np), np.std(arr_np), mean_tol=0.05
+        )
+
+
+@pytest.mark.xfail(reason="cuNumeric raises NotImplementedError")
+@pytest.mark.parametrize("size", (1024, 1025))
+def test_randint_bool(size):
+    """Test randint with boolean output dtype."""
+    arr_np, arr_num = gen_random_from_both("randint", 2, size=size, dtype=bool)
+    assert_distribution(
+        arr_num, np.mean(arr_np), np.std(arr_np), mean_tol=0.05
+    )
+    # NumPy pass
+    # cuNumeric LEGATE_TEST=1 or size > 1024:
+    # NotImplementedError: type for random.integers has to be int64 or int32
+    # or int16
+
+
+@pytest.mark.parametrize("low, high", LOW_HIGH, ids=str)
+@pytest.mark.parametrize("size", SIZES, ids=str)
+def test_random_integers(low, high, size):
+    arr_np, arr_num = gen_random_from_both(
+        "random_integers", low=low, high=high, size=size
+    )
+    assert arr_np.shape == arr_num.shape
+    # skip distribution assert for small arrays
+    if arr_np.size > 1024:
```

Comment:
I do generally think it is better to split up tests that have conditionals like this. Ideally tests only have one assertion, but in situations where there are multiple assertions, it's better for all cases in a given test to see the exact same set of assertions.

Reply:
I would agree in general, but for this test suite I'd say it's better to keep them together, the `SIZES` are meant to be tested for each of the APIs that accepts it, and combined with other inputs, I don't see much point pulling out the smaller ones and create a bunch of redundant code just because we don't have a better way to assert the distribution.

An alternative would be xfailing the tests on small arrays, and improve the distribution assertion later so the params eventually go through the same assert. I'm not sure if we're willing to xfail tests on test code issues though.

Reply:
> redundant code just because we don't have a better way to assert the distribution.

That's precisely a reason for redundant code (in tests), though. Tests should be as comprehensive and fine-grained as possible, whatever it takes. DRY is a top-level guiding design principle for the library code itself, but for test code DRY is only a nice-to-have, at best. 

I'm not going to press for a change but I will mention how I would organize this:
```python
SMALL_RNG_SIZES = [...]
LARGE_RNG_SIZES = [...]
ALL_RNG_SIZES = SMALL_RNG_SIZES + LARGE_RNG_SIZES

class TestRandomIntegers:

    @pytest.mark.parametrize("low, high", LOW_HIGH, ids=str)
    @pytest.mark.parametrize("size", ALL_RNG_SIZES, ids=str)
    def test_basic_stats(low, high, size):
        # shape, high, low respected, etc

    @pytest.mark.parametrize("low, high", LOW_HIGH, ids=str)
    @pytest.mark.parametrize("size", LARGE_RNG_SIZES, ids=str)
    def test_distribution(low, high, size):
        # test detailed distribution
```

And I would put the xfails in a separate test entirely (in this case).

As an aside, currently the `SIZES` list contains `xfail` params inside. I would normally advise to have "consts" like `SIZE` only contain bare data, and only utilize `xfail` etc inside the `parametrize` call itself. Yes, reworking to be that way could also end up with "redundancies", but what happens if/when a library update fixes *one* of the tests for that value, but not others? I think this kind of re-use is somewhat illusory, or at least sets up a situation that will likely require more work in the future.

Reply:
> but what happens if/when a library update fixes one of the tests for that value, but not others?

I was informed at one point that we want to keep the cases as XPASS after something is fixed, good to know that is not really the case.

On that note, which method would you prefer for handling the individual xfails?
Leaving the xfail ones outside the param list and extend in each case?
```
@pytest.mark.parametrize("size", ALL_RNG_SIZES + [pytest.param(-12345, marks=pytest.mark.xfail)], ids=str)
def test_basic_stats(size):
    ...

@pytest.mark.parametrize("size", ALL_RNG_SIZES + [-12345], ids=str)
def test_something_doesnt_fail(size):
    ...
```
Or keep a full param list and xfail inside the case?
```
@pytest.mark.parametrize("size", ALL_RNG_SIZES, ids=str)
def test_basic_stats(lsize):
    try:
        ...
    except:
        if size == -12345:
            pytest.xfail()
        raise
```

Reply:
I think it situational:

* if there is one (or a few cases) out of many that are failing, then the first version above (inline in the `parametrize`) seems good
* If there are lots of cases that are currently expected to fail, I think its probably cleaner to just have a dedicated test with a "global" `xfail` mark for the entire test